### PR TITLE
Close #984, Close #1282 Test PR on Release Branch (action)

### LINKED ISSUE #984: Recreate D7 "Test in Pantheon multidev" Bitbucket pipeline
## Motivation
It would be useful to be able to test PRs on actual Quickstart sites before merging.  We had the ability to do this with D7 Quickstart thanks to a custom Bitbucket Pipeline that @jwmoore1 and @trackleft put together.

## Proposed Resolution
Recreate D7 "Test in Pantheon multidev" Bitbucket pipeline as GitHub action (or similar) for Quickstart 2 repo.

### Describe the solution you'd like
When reviewing a Quickstart PR, I'd like to be able to manually trigger a GitHub actions workflow (or similar) that applies the PR changes to a multidev on a Pantheon site.

## Additional context
[D7 Bitbucket pipeline config](https://bitbucket.org/ua_drupal/ua_quickstart/src/99fa4c5ecc57d3726987081f9131e5a7777feb64/bitbucket-pipelines.yml#lines-49:79)
[D7 PR](https://bitbucket.org/ua_drupal/ua_quickstart/pull-requests/589)
[D7 Jira Issue](https://jira.arizona.edu/browse/UADIGITAL-2115)

### LINKED ISSUE #1282: Add CI workflows for testing PRs on release branches
## Motivation
We need to be able to test whether a change works with our release branches before merging.

## Proposed Resolution
Add CI workflows that are either automatically run against all of our supported release branches or can manually be run against specific branches.

## Additional context
GItHub action's [matrix job strategy feature](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstrategymatrix) is probably useful for this.

### PR DESCRIPTION
**Pull request by kevcooper**
## Related Issue
#1282 

## Types of changes

**Arizona Quickstart** (install profile, custom modules, custom theme)
- Patch release changes
   - [ ] Bug fix
   - [ ] Accessibility, performance, or security improvement
   - [ ] Critical institutional link or brand change
- Minor release changes
   - [x] New feature
   - [ ] Breaking or visual change to existing behavior
   - [ ] Non-critical brand change
   - [ ] New internal API or API improvement with backwards compatibility
   - [ ] Risky or disruptive cleanup to comply with coding standards
   - [ ] High-risk or disruptive change (requires upgrade path, risks regression, etc.)
- [ ] Other or unknown

**Drupal core**
- Patch release changes
   - [ ] Security update
   - [ ] Patch level release (non-security bug-fix release)
   - [ ] Patch removal that's no longer necessary
- Minor release changes
   - [ ] Major or minor level update
- [ ] Other or unknown

**Drupal contrib projects**
- Patch release changes
   - [ ] Security update
   - [ ] Patch or minor level update
   - [ ] Add new module
   - [ ] Patch removal that's no longer necessary
- Minor release changes
   - [ ] Major level update
- [ ] Other or unknown

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: .github/workflows/test-pr-on-release-branch.yml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        uses: actions/checkout@v2
+
+      - name: Setup PHP
+        uses: shivammathur/setup-php@d2f58713aaf7809d0c4d11e827c9e9dbbc55b34e
+        with:
+          php-version: '7.4'
+
+      - name: Apply PR diff to release branch
+        id: apply
+        run: |
```
Comment (joeparsons) [inline .github/workflows/test-pr-on-release-branch.yml:38]:
```suggestion
        run: |
          set -e
```

File: .github/workflows/test-pr-on-release-branch.yml [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+          set -x
+          git config --global user.email "${GITHUB_ACTOR}@users.noreply.github.com"
+          git config --global user.name "${GITHUB_ACTOR}"
+          git fetch
+          git checkout "${RELEASE_BRANCH}"
```
Comment (joeparsons) [inline .github/workflows/test-pr-on-release-branch.yml:43]:
```suggestion
          git checkout "${{ matrix.release }}"
```

File: .github/workflows/test-pr-on-release-branch.yml [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        run: |
+          set -x
+          git config --global user.email "${GITHUB_ACTOR}@users.noreply.github.com"
+          git config --global user.name "${GITHUB_ACTOR}"
+          git fetch
+          git checkout "${{ matrix.release }}"
+          curl https://patch-diff.githubusercontent.com/raw/az-digital/az_quickstart/pull/1459.patch | git apply --reject --whitespace=fix
+          
+      - name: Composer Install
+        working-directory: ./az_quickstart
```
Comment (joeparsons) [inline .github/workflows/test-pr-on-release-branch.yml:50]:
```suggestion
```

File: .github/workflows/test-pr-on-release-branch.yml [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        run: |
+          # Clone scaffolding repo into ./app
+          git clone --branch ${{ matrix.release }} https://github.com/az-digital/az-quickstart-scaffolding.git ./app
+          # Update az-digital/az_quickstart dependency to reference local (patched) copy of Quickstart
```
Comment (joeparsons) [inline .github/workflows/test-pr-on-release-branch.yml:54]:
```suggestion
          # Update az-digital/az_quickstart dependency to reference local (patched) copy of Quickstart
          cd app
```

File: .github/workflows/test-pr-on-release-branch.yml [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+          set -x
+          git config --global user.email "${GITHUB_ACTOR}@users.noreply.github.com"
+          git config --global user.name "${GITHUB_ACTOR}"
+          git fetch
+          git checkout "${{ matrix.release }}"
+          curl https://patch-diff.githubusercontent.com/raw/az-digital/az_quickstart/pull/1459.patch | git apply --reject --whitespace=fix
+          
+      - name: Composer Install
+        run: |
+          # Clone scaffolding repo into ./app
```
Comment (joeparsons) [inline .github/workflows/test-pr-on-release-branch.yml:51]:
```suggestion
          set -x
          # Clone scaffolding repo into ./app
```

File: .github/workflows/test-pr-on-release-branch.yml [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+        working-directory: ./az_quickstart
+        run: |
+          set -x
+          git config --global user.email "${GITHUB_ACTOR}@users.noreply.github.com"
+          git config --global user.name "${GITHUB_ACTOR}"
+          git fetch
+          git checkout "${{ matrix.release }}"
+          PR_NUMBER=$(jq --raw-output .pull_request.number "$GITHUB_EVENT_PATH")
+          echo "${PR_NUMBER}"
+          curl https://patch-diff.githubusercontent.com/raw/az-digital/az_quickstart/pull/1408.path | git apply --reject --whitespace=fix --allow-empty
```
Comment (trackleft) [inline .github/workflows/test-pr-on-release-branch.yml:64]:
```suggestion
          curl https://patch-diff.githubusercontent.com/raw/az-digital/az_quickstart/pull/1408.patch | git apply --reject --whitespace=fix --allow-empty
```

File: .github/workflows/test-pr-on-release-branch.yml [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+          # Clone scaffolding repo into ./app
+          git clone --branch ${{ matrix.release }} https://github.com/az-digital/az-quickstart-scaffolding.git ./app
+          # Update az-digital/az_quickstart dependency to reference local (patched) copy of Quickstart
+          cd app
+          composer config repositories.localqs path ${GITHUB_WORKSPACE}/az_quickstart
+          composer require --no-update az-digital/az_quickstart:\*@dev
+          composer install         
+       
+      - name: Install Drush
+        run: php ./app/vendor/bin/drush site:install az_quickstart --account-pass=admin --db-url=mysql://root:@127.0.0.1:${{ job.services.mysql.ports[3306] }}/db --yes
```
Comment (joeparsons) [inline .github/workflows/test-pr-on-release-branch.yml:81]:
```suggestion
        run: php ./app/vendor/bin/drush site:install az_quickstart --account-pass=admin --db-url=mysql://root:@127.0.0.1:${{ job.services.mysql.ports[3306] }}/db --yes --root=./app/web
```

File: .github/workflows/test-pr-on-release-branch.yml [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+          git fetch
+          git checkout "${{ matrix.release }}"
+          PR_NUMBER=$(jq --raw-output .pull_request.number "$GITHUB_EVENT_PATH")
+          echo "${PR_NUMBER}"
+          curl https://patch-diff.githubusercontent.com/raw/az-digital/az_quickstart/pull/1408.diff | git apply --reject --whitespace=fix --allow-empty
+          
+          
+      - name: Composer Install
+        id: install
+        if: steps.apply.outcome == 'failure'
```
Comment (joeparsons) [inline .github/workflows/test-pr-on-release-branch.yml:69]:
```suggestion
        if: steps.apply.outcome == 'success'
```

File: .github/workflows/test-pr-on-release-branch.yml [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+          set -x
+          # Clone scaffolding repo into ./app
+          git clone --branch ${{ matrix.release }} https://github.com/az-digital/az-quickstart-scaffolding.git ./app
+          # Update az-digital/az_quickstart dependency to reference local (patched) copy of Quickstart
+          cd app
+          composer config repositories.localqs path ${GITHUB_WORKSPACE}/az_quickstart
+          composer require --no-update az-digital/az_quickstart:\*@dev
+          composer install         
+       
+      - name: Install Drush
```
Comment (joeparsons) [inline .github/workflows/test-pr-on-release-branch.yml:80]:
```suggestion
      - name: Install Quickstart with Drush
```

File: .github/workflows/test-pr-on-release-branch.yml [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+        run: php ./app/vendor/bin/drush site:install az_quickstart --account-pass=admin --db-url=mysql://root:@127.0.0.1:${{ job.services.mysql.ports[3306] }}/db --yes
+
+      - name: PHP Unit Tests
+        run: |
+          # ./app/vendor/bin/phpunit --configuration web/profiles/custom/az_quickstart/phpunit.xml.dist --bootstrap web/core/tests/bootstrap.php web/profiles/custom/az_quickstart
+          set -x
+          sudo mkdir -p /var/www/html/sites/simpletest
```
Comment (joeparsons) [inline .github/workflows/test-pr-on-release-branch.yml:87]:
```suggestion
          php -S 127.0.0.1:8080 -t web &
          sudo mkdir -p /var/www/html/sites/simpletest
```

File: .github/workflows/test-pr-on-release-branch.yml [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+          sudo chown www-data:www-data -R /var/www/html/sites/simpletest
+          sudo mkdir -p /var/www/html/sites/default/files/simpletest
+          sudo chown www-data:www-data -R /var/www/html/sites/default/files/simpletest
+          sudo CHROMEDRIVER_VERSION=LATEST_$(/usr/bin/google-chrome --product-version | cut -d . -f 1-3) npm install --no-progress --unsafe-perm --global chromedriver
+          sudo apt-get install daemonize
+          export SIMPLETEST_BASE_URL="http://localhost"
```
Comment (joeparsons) [inline .github/workflows/test-pr-on-release-branch.yml:93]:
```suggestion
          export SIMPLETEST_BASE_URL="http://127.0.0.1:8080"
```

File: .github/workflows/test-pr-on-release-branch.yml [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+      - name: PHP Unit Tests
+        run: |
+          # ./app/vendor/bin/phpunit --configuration web/profiles/custom/az_quickstart/phpunit.xml.dist --bootstrap web/core/tests/bootstrap.php web/profiles/custom/az_quickstart
+          set -x
+          sudo mkdir -p /var/www/html/sites/simpletest
+          sudo chown www-data:www-data -R /var/www/html/sites/simpletest
+          sudo mkdir -p /var/www/html/sites/default/files/simpletest
+          sudo chown www-data:www-data -R /var/www/html/sites/default/files/simpletest
+          sudo CHROMEDRIVER_VERSION=LATEST_$(/usr/bin/google-chrome --product-version | cut -d . -f 1-3) npm install --no-progress --unsafe-perm --global chromedriver
+          sudo apt-get install daemonize
```
Comment (joeparsons) [inline .github/workflows/test-pr-on-release-branch.yml:92]:
```suggestion
```

File: .github/workflows/test-pr-on-release-branch.yml [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+      - name: PHP Unit Tests
+        run: |
+          # ./app/vendor/bin/phpunit --configuration web/profiles/custom/az_quickstart/phpunit.xml.dist --bootstrap web/core/tests/bootstrap.php web/profiles/custom/az_quickstart
+          set -x
+          php -S 127.0.0.1:8080 -t web &
+          sudo mkdir -p /var/www/html/sites/simpletest
+          sudo chown www-data:www-data -R /var/www/html/sites/simpletest
+          sudo mkdir -p /var/www/html/sites/default/files/simpletest
+          sudo chown www-data:www-data -R /var/www/html/sites/default/files/simpletest
+          export SIMPLETEST_BASE_URL="http://127.0.0.1:8080"
```
Comment (joeparsons) [inline .github/workflows/test-pr-on-release-branch.yml:92]:
```suggestion
          export SIMPLETEST_DB="mysql://root:@127.0.0.1:${{ job.services.mysql.ports[3306] }}/db"
          export SIMPLETEST_BASE_URL="http://127.0.0.1:8080"
```

File: .github/workflows/test-pr-on-release-branch.yml [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+          # php -S 127.0.0.1:8080 -t web &
+          php -S http://localhost:8080 -t web &
+          sudo mkdir -p /var/www/html/sites/simpletest
+          sudo chown www-data:www-data -R /var/www/html/sites/simpletest
+          sudo mkdir -p /var/www/html/sites/default/files/simpletest
+          sudo chown www-data:www-data -R /var/www/html/sites/default/files/simpletest
+          # export SIMPLETEST_BASE_URL="http://127.0.0.1:8080"
+          export SIMPLETEST_BASE_URL="http://localhost:8080"
+          # export MINK_DRIVER_ARGS_WEBDRIVER='["chrome", {"browserName":"chrome","chromeOptions":{"args":["--disable-gpu","--headless", "--no-sandbox"]}}, "http://127.0.0.1:${{ job.services.chromedriver.ports[9515] }}"]'
+          export MINK_DRIVER_ARGS_WEBDRIVER='["chrome", {"browserName":"chrome","chromeOptions":{"args":["--disable-gpu","--headless", "--no-sandbox"]}}, "http://localhost:${{ job.services.chromedriver.ports[9515] }}"]'
```
Comment (joeparsons) [inline .github/workflows/test-pr-on-release-branch.yml:96]:
```suggestion
          # export SIMPLETEST_BASE_URL="http://127.0.0.1:8080"
          export SIMPLETEST_BASE_URL="http://localhost:8080"
          export SIMPLETEST_DB="mysql://root:@127.0.0.1:${{ job.services.mysql.ports[3306] }}/db"
          export MINK_DRIVER_ARGS_WEBDRIVER='["chrome", {"browserName":"chrome","chromeOptions":{"args":["--disable-gpu","--headless", "--no-sandbox"]}}, "http://127.0.0.1:${{ job.services.chromedriver.ports[9515] }}"]'
          # export MINK_DRIVER_ARGS_WEBDRIVER='["chrome", {"browserName":"chrome","chromeOptions":{"args":["--disable-gpu","--headless", "--no-sandbox"]}}, "http://localhost:${{ job.services.chromedriver.ports[9515] }}"]'
```

File: .github/workflows/test-pr-on-release-branch.yml [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+       
+      - name: Install Drush
+        run: php ./app/vendor/bin/drush site:install az_quickstart --account-pass=admin --db-url=mysql://root:@127.0.0.1:${{ job.services.mysql.ports[3306] }}/db --yes
+
+      - name: PHP Unit Tests
+        run: |
+          # ./app/vendor/bin/phpunit --configuration web/profiles/custom/az_quickstart/phpunit.xml.dist --bootstrap web/core/tests/bootstrap.php web/profiles/custom/az_quickstart
+          set -x
+          # php -S 127.0.0.1:8080 -t web &
+          php -S http://localhost:8080 -t web &
```
Comment (joeparsons) [inline .github/workflows/test-pr-on-release-branch.yml:88]:
```suggestion
          php -S 127.0.0.1:8080 -t web &
          # php -S http://localhost:8080 -t web &
```

File: .github/workflows/test-pr-on-release-branch.yml [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+          sudo mkdir -p /var/www/html/sites/simpletest
+          sudo chown www-data:www-data -R /var/www/html/sites/simpletest
+          sudo mkdir -p /var/www/html/sites/default/files/simpletest
+          sudo chown www-data:www-data -R /var/www/html/sites/default/files/simpletest
+          # export SIMPLETEST_BASE_URL="http://127.0.0.1:8080"
+          export SIMPLETEST_BASE_URL="http://localhost:8080"
```
Comment (joeparsons) [inline .github/workflows/test-pr-on-release-branch.yml:94]:
```suggestion
          export SIMPLETEST_BASE_URL="http://127.0.0.1:8080"
          # export SIMPLETEST_BASE_URL="http://localhost:8080"
```

File: .github/workflows/test-pr-on-release-branch.yml [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+          cd app
+          composer config repositories.localqs path ${GITHUB_WORKSPACE}/az_quickstart
+          composer require --no-update az-digital/az_quickstart:\*@dev
+          composer install         
+       
+      - name: Install Drush
+        run: php ./app/vendor/bin/drush site:install az_quickstart --account-pass=admin --db-url=mysql://root:@127.0.0.1:${{ job.services.mysql.ports[3306] }}/db --yes
+
+      - name: PHP Unit Tests
+        run: |
```
Comment (joeparsons) [inline .github/workflows/test-pr-on-release-branch.yml:84]:
```suggestion
        working_directory: ./app
        run: |
```

File: .github/workflows/test-pr-on-release-branch.yml [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+          # ./app/vendor/bin/phpunit --configuration web/profiles/custom/az_quickstart/phpunit.xml.dist --bootstrap web/core/tests/bootstrap.php web/profiles/custom/az_quickstart
+          set -x
+          php -S 127.0.0.1:8080 -t web &
+          # php -S http://localhost:8080 -t web &
+          sudo mkdir -p /var/www/html/sites/simpletest
+          sudo chown www-data:www-data -R /var/www/html/sites/simpletest
```
Comment (joeparsons) [inline .github/workflows/test-pr-on-release-branch.yml:90]:
```suggestion
          mkdir -p web/sites/simpletest
          chmod 777  web/sites/simpletest
```

File: .github/workflows/test-pr-on-release-branch.yml [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+          sudo mkdir -p /var/www/html/sites/default/files/simpletest
+          sudo chown www-data:www-data -R /var/www/html/sites/default/files/simpletest
```
Comment (joeparsons) [inline .github/workflows/test-pr-on-release-branch.yml:92]:
```suggestion
          # sudo mkdir -p /var/www/html/sites/default/files/simpletest
          # sudo chown www-data:www-data -R /var/www/html/sites/default/files/simpletest
```

File: .github/workflows/test-pr-on-release-branch.yml [hunk 20]
```diff
@@ -0,0 +1,119 @@
...
+          export SIMPLETEST_BASE_URL="http://127.0.0.1:8080"
+          # export SIMPLETEST_BASE_URL="http://localhost:8080"
+          export SIMPLETEST_DB="mysql://root:@127.0.0.1:${{ job.services.mysql.ports[3306] }}/db"
+          export MINK_DRIVER_ARGS_WEBDRIVER='["chrome", {"browserName":"chrome","chromeOptions":{"args":["--disable-gpu","--headless", "--no-sandbox"]}}, "http://127.0.0.1:${{ job.services.chromedriver.ports[9515] }}"]'
+          # export MINK_DRIVER_ARGS_WEBDRIVER='["chrome", {"browserName":"chrome","chromeOptions":{"args":["--disable-gpu","--headless", "--no-sandbox"]}}, "http://localhost:${{ job.services.chromedriver.ports[9515] }}"]'
+          find ./app/web/profiles/custom/az_quickstart/ -name "*Test.php" | ./app/vendor/bin/fastest -vvv "./app/vendor/bin/phpunit {} --configuration ./app/web/profiles/custom/az_quickstart/phpunit.xml.dist --bootstrap ./app/web/core/tests/bootstrap.php ./app/web/profiles/custom/az_quickstart;"
```
Comment (joeparsons) [inline .github/workflows/test-pr-on-release-branch.yml:98]:
```suggestion
          find ./web/profiles/custom/az_quickstart/ -name "*Test.php" | ./vendor/bin/fastest -vvv "./vendor/bin/phpunit {} --configuration ./web/profiles/custom/az_quickstart/phpunit.xml.dist --bootstrap ./web/core/tests/bootstrap.php ./web/profiles/custom/az_quickstart;"
```

**Comment by joeparsons** [comment]
Closing this for now.  May reopen if needed for https://github.com/az-digital/az_quickstart/issues/1282

**Pull request closed without merging**